BPOR/BPOR.Rms/Controllers/NotifyCallbackController.cs

```diff
@@ -0,0 +1,104 @@
+using BPOR.Domain.Entities;
+using BPOR.Registration.Stream.Handler.Services;
+using BPOR.Rms.Constants;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Options;
+using NIHR.NotificationService.Settings;
+
+namespace BPOR.Rms.Controllers;
+
+[AllowAnonymous]
+[ApiController]
+[Route("NotifyCallback")]
+public class NotifyCallbackController(
+    ParticipantDbContext context,
+    IOptions<NotificationServiceSettings> settings,
+    IRefDataService refDataService
+) : ControllerBase
+{
+    [HttpPost]
+    [Route("ReceiveCallback")]
+    public async Task<IActionResult> ReceiveCallback([FromBody] NotifyCallbackMessage message,
+        CancellationToken cancellationToken)
+    {
+        var token = Request.Headers.Authorization.ToString().Replace("Bearer ", "");
+        if (token != settings.Value.BearerToken)
+        {
+            return Unauthorized();
+        }
+
+        if (message.Reference == "PreviewEmailReference")
+        {
+            return Ok();
+        }
+
+        if (!int.TryParse(message.Reference, out var participantEmailId))
+        {
+            return BadRequest("Invalid reference.");
+        }
+
+        var participantEmail = await context.EmailCampaignParticipants
+            .FindAsync(new object[] { participantEmailId }, cancellationToken);
+
+        if (participantEmail == null)
+        {
+            return NotFound();
+        }
+
+        switch (message.Status)
+        {
+            case "delivered":
+                participantEmail.DeliveredAt = DateTime.UtcNow;
+                participantEmail.DeliveryStatusId =
+                    refDataService.GetEmailDeliveryStatusId(EmailDeliveryStatus.Delivered);
+                break;
+            case "permanent-failure":
+            case "technical-failure":
+                participantEmail.DeliveryStatusId = refDataService.GetEmailDeliveryStatusId(EmailDeliveryStatus.Failed);
+                break;
+            default:
+                return BadRequest("Invalid status.");
+        }
+
+        await context.SaveChangesAsync(cancellationToken);
+
+        return Ok();
+    }
+
+    [HttpGet]
+    [Route("registerinterest")]
+    public async Task<IActionResult> RegisterInterest([FromQuery] string reference, CancellationToken cancellationToken)
+    {
+        var participant = await context.EmailCampaignParticipants
```

Comment:
We don't want to expose the reference directly here.
An encrypted string would be preferable, so it is hidden, but we can still look it up directly in the database after decryption.